src/blas/impl/KokkosBlas1_dot_mv_impl.hpp

```diff
@@ -44,617 +44,125 @@
-template <class RV, class XMV, class YMV, class SizeType,
-          const int XMV_rank = XMV::rank, const int YMV_rank = YMV::rank>
-struct MV_V_Dot_Invoke_Impl {
-  static void run(const RV& r, const XMV& X, const YMV& Y,
-                  const SizeType numRows);
-};
-
-template <class RV, class XMV, class YMV, class SizeType>
-struct MV_V_Dot_Invoke_Impl<RV, XMV, YMV, SizeType, 2, 1> {
-  static void run(const RV& r, const XMV& X, const YMV& Y,
-                  const SizeType numRows) {
-    static_assert(
-        static_cast<int>(XMV::rank) == 2 && static_cast<int>(YMV::rank) == 1,
-        "XMV must have rank 2, and YMV must have rank 1.");
-    static_assert(static_cast<int>(RV::rank) == 1, "Rank of r must be 1.");
-    static_assert(std::is_integral<SizeType>::value,
-                  "SizeType must be a built-in integer type.");
-
-    typedef typename XMV::execution_space execution_space;
-    typedef SizeType size_type;
-    typedef Kokkos::RangePolicy<execution_space, size_type> range_type;
-
-    typedef MV_V_Dot_Functor<RV, XMV, YMV, size_type> op_type;
-    constexpr bool reverseOrder = false;
-    op_type op(r, X, Y, reverseOrder);
-    Kokkos::parallel_reduce("KokkosBlas::Dot::2_1", range_type(0, numRows), op,
-                            r);
-  }
-};
-
-template <class RV, class XMV, class YMV, class SizeType>
-struct MV_V_Dot_Invoke_Impl<RV, XMV, YMV, SizeType, 1, 2> {
-  static void run(const RV& r, const XMV& X, const YMV& Y,
-                  const SizeType numRows) {
-    static_assert(
-        static_cast<int>(XMV::rank) == 1 && static_cast<int>(YMV::rank) == 2,
-        "XMV must have rank 1, and YMV must have rank 2.");
-    static_assert(static_cast<int>(RV::rank) == 1, "Rank of r must be 1.");
-    static_assert(std::is_integral<SizeType>::value,
-                  "SizeType must be a built-in integer type.");
-
-    typedef typename XMV::execution_space execution_space;
-    typedef SizeType size_type;
-    typedef Kokkos::RangePolicy<execution_space, size_type> range_type;
-
-    // Use the "reverse arguments" mode of the functor.
-    typedef MV_V_Dot_Functor<RV, YMV, XMV, size_type> op_type;
-    constexpr bool reverseOrder = true;
-    op_type op(r, Y, X, reverseOrder);
-    Kokkos::parallel_reduce("KokkosBlas::Dot::1_2", range_type(0, numRows), op,
-                            r);
+  void operator()(const TeamMem& t) const {
+    size_type globalRank = t.league_rank();
+    size_type localRank  = globalRank % teamsPerDot;
+    size_type i          = globalRank / teamsPerDot;
+    size_type xcol       = x.extent(1) == 1 ? 0 : i;
+    size_type ycol       = y.extent(1) == 1 ? 0 : i;
+    size_type length     = x.extent(0);
+
+    dot_type localResult = KAT::zero();
+    size_type baseInd    = chunkSize * localRank;
+    Kokkos::parallel_reduce(
+        Kokkos::TeamThreadRange(t, chunkSize),
+        [&](size_type k, dot_type& update) {
+          if (baseInd + k < length) {
```

Comment:
Consider evaluating performance without this if statement in the `parallel_reduce`. For BatchedGemm, we have a "full dot product" loop followed by a "partial dot product" loop. The "partial dot product" loop either computes the last full dot product or the last partial dot product based on bounds checking done in the constructor.

Reply:
That is a good point. I'm thinking we should do some better load balancing:
```
size_type chunkBegin = chunk * length / chunkSize
size_type chunkEnd = (chunk + 1) * length / chunkSize
Kokkos::TeamThreadRange(t, chunkBegin, chunkEnd), ...
```
so that the last team in a column doesn't get less work than the others, and also to eliminate the ``if`` inside the loop.